### ISSUE TITLE
self.coral_bias explanation

### ISSUE THREAD

**Issue by 421psh** [Raschka-research-group/coral_pytorch#17]
Hi!

Now I'm adding coral loss to my current project and I don't understand how `self.coral_bias` works. During loss calculation we are comparing levels and logits like in your example: 
`levels = torch.tensor([[1., 1., 0., 0.], [1., 0., 0., 0.], [1., 1., 1., 1.]])`
`logits = torch.tensor([[2.1, 1.8, -2.1, -1.8], [1.9, -1., -1.5, -1.3], [1.9, 1.8, 1.7, 1.6]])`

But `self.coral_bias = torch.nn.Parameter(torch.zeros(num_classes-1).float())` makes all logits to look like:
`logits = torch.tensor([[2.1, 2.1, 2.1, 2.1], [1.9, 1.9, 1.9, 1.9], [1.7, 1.7, 1.7, 1.7]])` where all elements inside tensor have same values.
So after sigmoid and probability estimation tensor could be only in two states: `[0., 0., 0., 0.]` or `[1., 1., 1., 1.,]`. Or maybe I'm not right and do not understand something. Could you give an explanation to this, please?

**Comment by rasbt** [comment]
I see your concern! So, you can think of CORAL as follows

```python
import torch


num_classes = 10
x = torch.rand(1024, 512)

output_layer = torch.nn.Linear(512, (num_classes-1), bias=False)
output_biases = torch.nn.Parameter(torch.zeros(num_classes-1).float())

logits = output_layer(x) + output_biases
```

Above, `x` is some input data. If all the input data is the same (e.g., `x = torch.ones(1024, 512)` ), then you might think that all the outputs are the same, like

```python
tensor([[-0.2068, -0.3038, -0.3499,  ...,  0.1888,  0.0703,  0.4388],
        [-0.2068, -0.3038, -0.3499,  ...,  0.1888,  0.0703,  0.4388],
        [-0.2068, -0.3038, -0.3499,  ...,  0.1888,  0.0703,  0.4388],
        ...,
        [-0.2068, -0.3038, -0.3499,  ...,  0.1888,  0.0703,  0.4388],
        [-0.2068, -0.3038, -0.3499,  ...,  0.1888,  0.0703,  0.4388],
        [-0.2068, -0.3038, -0.3499,  ...,  0.1888,  0.0703,  0.4388]],
       grad_fn=<MmBackward0>)
```

But you have to keep in mind that the biases will change during training. E.g., think of changing

```python
output_biases = torch.nn.Parameter(torch.zeros(num_classes-1).float())
```

to

```python
output_biases = torch.nn.Parameter(torch.rand(num_classes-1).float())
```
to see the effect.

**Comment by 421psh** [comment]
Unfortunately, I still can't wrap my head around it. Now, I don't understand two things:

1. In your example, in Linear layer out_features = (num_classes-1), but in code Linear layer is `self.coral_weights = torch.nn.Linear(self.size_in, 1, bias=False)`, i.e. out_features = 1. What is right?
2. During training, biases change to fit the data. This is indisputable. But here, biases are always zeros. Are they trainable?

I apologize in advance if I do not understand some basic things and you have to explain it to me. For me, this point is really not obvious.

**Comment by rasbt** [comment]
> 1. In your example, in Linear layer out_features = (num_classes-1), but in code Linear layer is self.coral_weights = torch.nn.Linear(self.size_in, 1, bias=False), i.e. out_features = 1. What is right?

Here 

    self.coral_weights = torch.nn.Linear(self.size_in, 1, bias=False)

is due to the weight sharing, so there is 1 shared weight. But the number of outputs is `num_classes-1` because of the biases that you add. 

Think of it as

```python
self.output_layer = torch.nn.Linear(num_hidden, 1, bias=False)
self.output_biases = torch.nn.Parameter(
            torch.zeros(NUM_CLASSES-1).float())

logits = self.output_layer(x) + self.output_biases.view(1, -1)
```

> 2. During training, biases change to fit the data. This is indisputable. But here, biases are always zeros. Are they trainable?

Yes they are trainable due to using `torch.nn.Parameter` rather than `torch.tensor`. 0's are just the starting biases (you can also initialize them to small random values; we found that this sometimes speeds up training/convergence but is not necessary).

Let me know if you have any other concerns!
